### PR TITLE
Support doctrine/mongodb-odm ^2.x & doctrine/mongodb-odm-bundle ^4.x

### PR DESCRIPTION
**Pull request by romainnorberg**
- support both version of `doctrine/mongodb-odm` (`^1.0 || ^2.0`)
    - https://github.com/doctrine/mongodb-odm/blob/master/UPGRADE-2.0.md
    - **BC**: https://github.com/doctrine/mongodb-odm/pull/2069
- support both version of  `doctrine/mongodb-odm-bundle` (`^3.0 || ^4.0`) 
    - https://github.com/doctrine/DoctrineMongoDBBundle/blob/master/UPGRADE-4.0.md
- run php-cs fixer _(lint CI build)_
- updated `FilterTypeGuesserTest`

I am targeting this branch, because this is not BC.

### CODE REVIEW

**Comment by core23** [comment]
AFAIK you could do this on the stable branch. 

WDYT @sonata-project/contributors

File: src/Guesser/FilterTypeGuesser.php [hunk 1]
```diff
@@ -49,10 +49,10 @@ public function guessType($class, $property, ModelManagerInterface $modelManager
             $mapping = $metadata->fieldMappings[$propertyName];
 
             switch ($mapping['type']) {
-                case ClassMetadataInfo::ONE:
-                case ClassMetadataInfo::MANY:
-                    //case ClassMetadataInfo::MANY_TO_ONE:
-                    //case ClassMetadataInfo::MANY_TO_MANY:
+                case ClassMetadata::ONE:
+                case ClassMetadata::MANY:
+                    //case ClassMetadata::MANY_TO_ONE:
```
Comment (core23) [inline src/Guesser/FilterTypeGuesser.php]: This comments are superfluous, you can remove them

**Comment by OskarStark** [comment]
Yes let’s do it on the stable branch

**Comment by romainnorberg** [comment]
hello @core23 @OskarStark Thanks for reviews 👍 

I don't have a lot of experience in BC management _(I thought I had to target the master branch with these changes)_ and I am a novice in the contribution to open-source packages 😕 

I'll check why CI fail

**Comment by aturki** [comment]
Hello,
Both doctrine/mongodb-odm and doctrine/mongodb-odm-bundle require PHP version 7.2 or higher. This is a BC break with the current version 3.x. Shouldn't we target the master branch instead ?

**Comment by phansys** [comment]
IMO, from a strict semver point of view, if the changes are respecting BC we could accept them regardless their dependencies; since PHP is in fact a dependency itself.
But we have some [extra rules](https://github.com/sonata-project/SonataDoctrineMongoDBAdminBundle/blob/3.x/CONTRIBUTING.md#dependency-changes) around it which disallow the drop of a security maintained PHP version on a stable branch.

File: src/Builder/FormContractor.php [hunk 2]
```diff
@@ -13,7 +13,7 @@
 
 namespace Sonata\DoctrineMongoDBAdminBundle\Builder;
 
-use Doctrine\ODM\MongoDB\Mapping\ClassMetadataInfo;
+use Doctrine\ODM\MongoDB\Mapping\ClassMetadata;
```
Comment (core23) [inline src/Builder/FormContractor.php]: This class exists in odm 1.x and 2.x. You could support both version inside the composer.json file

**Comment by romainnorberg** [comment]
hello @core23 thanks for review.

- I updated composer.json file to support both version of mongodb-odm
- run php-cs fixer _(lint CI build)_
- updated `FilterTypeGuesserTest` -> i don't know if it's the right solution 👀

File: composer.json [hunk 3]
```diff
@@ -26,8 +26,8 @@
     ],
     "require": {
         "php": "^7.1",
-        "doctrine/mongodb-odm": "^1.0",
-        "doctrine/mongodb-odm-bundle": "^3.0",
+        "doctrine/mongodb-odm": "^1.0 || ^2.0",
+        "doctrine/mongodb-odm-bundle": "^4.0",
```
Comment (greg0ire) [inline composer.json]: Why are we not using `^3.0 || ^4.0` here? No way the 7.1 build works otherwise…

Comment (romainnorberg) [reply composer.json]: hello @greg0ire you're right, I updated the composer.json. thank u for the review

File: tests/Guesser/FilterTypeGuesserTest.php [hunk 4]
```diff
@@ -35,6 +35,9 @@ public function testThrowsOnMissingField()
 
         $class = 'My\Model';
         $property = 'whatever';
+
+        $this->metadata->hasAssociation($property)->willReturn(false);
+
```
Comment (greg0ire) [inline tests/Guesser/FilterTypeGuesserTest.php]: Why did you add this line exactly?

Comment (romainnorberg) [reply tests/Guesser/FilterTypeGuesserTest.php]: method `metadata->hasAssociation` must return a boolean _(typed since https://github.com/doctrine/mongodb-odm/pull/1846 I think)_ Without this line, test fail.


Test log https://travis-ci.org/sonata-project/SonataDoctrineMongoDBAdminBundle/jobs/593439895#L498

<details>
1) Sonata\DoctrineMongoDBAdminBundle\Tests\Guesser\FilterTypeGuesserTest::testThrowsOnMissingField
Failed asserting that exception of type "TypeError" matches expected exception "Sonata\DoctrineMongoDBAdminBundle\Model\MissingPropertyMetadataException". Message was: "Return value of Double\Doctrine\ODM\MongoDB\Mapping\ClassMetadata\P8::hasAssociation() must be of the type bool, null returned" at
/home/travis/build/sonata-project/SonataDoctrineMongoDBAdminBundle/src/Guesser/FilterTypeGuesser.php:48
/home/travis/build/sonata-project/SonataDoctrineMongoDBAdminBundle/tests/Guesser/FilterTypeGuesserTest.php:43
</details>

cf:
- https://github.com/romainnorberg/SonataDoctrineMongoDBAdminBundle/blob/master/src/Guesser/FilterTypeGuesser.php#L48

Comment (greg0ire) [reply tests/Guesser/FilterTypeGuesserTest.php]: Oh ok, there was probably no type declaration before in `ClassMetadata`, so it was fine with returning `null`. I think your fix is ok 👍

**Comment by romainnorberg** [comment]
thanks @all for your time & reviews

**Comment by OskarStark** [comment]
We thank you for this contribution, keep on going 👍🏻👍🏻🥳🥳
